bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/server/CallbackServer.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    public String getAuthorizationUrl() {
+        try {
+            return oacs.getAuthorizationUrl(callbackUrl, null, null);
+        } catch (OAuthException e) {
+            LOGGER.error("Error creating Authorization URL {}", e.getMessage());
+            return Constants.EMPTY;
+        }
+    }
+
+    public String getScope() {
+        return config.getScope();
+    }
+
+    public void start() {
+        LOGGER.debug("Start Callback Server for port {}", config.callbackPort);
+        if (!server.isEmpty()) {
+            LOGGER.debug("Callback server for port {} already started", config.callbackPort);
+            return;
+        }
+        server = Optional.of(new Server());
+        ServerConnector connector = new ServerConnector(server.get());
+        connector.setPort(config.callbackPort);
+        server.get().setConnectors(new Connector[] { connector });
+        ServletHandler servletHandler = new ServletHandler();
+        server.get().setHandler(servletHandler);
+        servletHandler.addServletWithMapping(CallbackServlet.class, Constants.CALLBACK_ENDPOINT);
+        try {
+            server.get().start();
+        } catch (Exception e) {
+            LOGGER.warn("Cannot start Callback Server for port {}, Error {}", config.callbackPort, e.getMessage());
+        }
+    }
+
+    public void stop() {
+        LOGGER.debug("Stop Callback Server");
+        try {
+            if (!server.isEmpty()) {
+                server.get().stop();
+                server = Optional.empty();
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Cannot start Callback Server for port {}, Error {}", config.callbackPort, e.getMessage());
+        }
+    }
+
+    public String getToken() {
+        if (token.isEmpty()) {
+            LOGGER.debug("Token empty - Manual Authorization needed at {}", callbackUrl);
+            return Constants.EMPTY;
+        } else {
+            if (token.get().isExpired(LocalDateTime.now(), 10)) {
+                LOGGER.trace("Token expired - start refreshing");
+                try {
+                    AccessTokenResponse act = oacs.refreshToken();
+                    token = Optional.of(act);
+                    listener.onAccessTokenResponse(act);
+                } catch (OAuthException | IOException | OAuthResponseException e) {
+                    LOGGER.warn("Error refreshing token. Reason: {} Manual Authorization needed at {}", e.getMessage(),
+                            callbackUrl);
+                }
+            } else {
+                LOGGER.trace("Token valid - do nothing");
+            }
+        }
+        return token.get().getAccessToken();
+    }
+
+    /**
+     * Function is used to restore last tokenResponse from Persistence.
+     * Use case e.g. is startup
+     *
+     * @param accessTokenResponse
+     */
+    public void setToken(AccessTokenResponse accessTokenResponse) {
+        token = Optional.of(accessTokenResponse);
+    }
```

Comment:
If this is only intended to be set once on startup, perhaps it should be a constructor parameter? The `keyMigration()` method could then create the object on its own, as I assume this is a rare scenario?

Reply:
Removed after rework. Also removed `keyMigration` which was introduced for convenience. If key migration doesn't work user will be asked with i18n status for manual authorization